%\VignetteIndexEntry{shinyMethyl}
\documentclass{article}
\title{shinyMethyl: a tool for interactive visualization of Illumina's 450k methylation arrays }
\author{Jean-Philippe Fortin}

\begin{document}
\SweaveOpts{concordance=TRUE}


\maketitle{}

\section{Introduction}
\texttt{shinyMethyl} is an interactive visualization tool for Illumina's 450k methylation arrays. The goal of \texttt{shinyMethyl} is twofold: 1) create a portable summary file of the methylation data that can be easily moved to a personal laptop and 2) launch an interactive visualization tool directly from the summarized data. The newly released \texttt{shiny} package \cite{shiny} makes the latter possible.
\\
\\
The extraction step uses the package \texttt{minfi} \cite{minfi} to read the raw data from the IDAT files and to handle \texttt{RGChannelSet} objects. With \texttt{shinyMethyl}, big data that cannot be loaded entirely into memory is not a problem; the function \texttt{extractFromTargets450k} can extract data subset by subset, with no more than 10 samples at a time. It is then possible to store all the raw data on a personal laptop, extract the data using \texttt{extractFromTargets450k} and launch \texttt{shinyMethyl} without the use of any external computer. This two-step approach makes the quality control much easier and enjoyable.

\section{\texttt{shinyMethyl} at a glance }
\texttt{shinyMethyl} contains four visualization panels: quality control, array design, gender clustering and principal component analysis (PCA). The first one is for exploration of the Beta-value (or M-value) distribution densities along with different control probes intensities. The second panel shows the study design of the arrays, colored by the user-chosen phenotype. The third panel implements the gender prediction developed in \texttt{minfi} and give the user the chance to choose a clustering cutoff and save the predicted gender. Finally, the last panel contains plots of different principal components colored by the user-chosen phenotype.
\subsection*{Example TCGA dataset}
The \texttt{shinyMethylData} package contains the extracted data of 369 Head and Neck cancer samples from The Cancer Genome Atlas (TCGA) \cite{toga}: 310 tumor samples, 50 matched normals and 9 replicates of a control cell lines. To run a \texttt{shinyMethyl} session with these samples:
<<shinyMethylData, eval=FALSE>>=
library(shinyMethylData)
runExampleTCGA()
@
\section{Extraction of the data}
In the current version of \texttt{shinyMethyl}, there are two ways to extract the necessary data before launching \texttt{shinyMethyl}. The extraction can be either from a previously constructed \texttt{RGChannelSet}, which is a class object from the package \texttt{minfi} that stores the raw red and green channel intensities, or directly from the IDAT files via a targets \texttt{data.frame} specifying the location of the files and other covariates of interest. 
\subsection{Extraction from a targets file}
The function \texttt{extractFromTargets450k} uses as input the same targets \texttt{data.frame} that is used in \texttt{minfi} to read IDAT files into R (into a \texttt{RGChannelSet}). The targets file is simply a \texttt{data.frame} containing information about the samples, such as phenotype and files location. It is assumed to have column named \texttt{Basename} pointing to the base name of a two color IDAT file, i.e. a name that can be made into a real IDAT file by appending either \texttt{\_Red.idat} or \texttt{\_Grn.idat}. For instance, \texttt{myFolder/880384095\_R01C02} is a basename for the pair of files \texttt{myFolder/880384095\_R01C02\_Grn.idat} and  \texttt{myFolder/880384095\_R01C02\_Red.idat}.For more information about how to create a targets file from a sample sheet, see the function \texttt{read.450k.sheet} in \texttt{minfi}
\\
\\
We will construct a targets file for the 6 samples included in the \texttt{minfiData} package: 
<<targetsConstruction, eval=TRUE>>=
library(shinyMethyl)
library(minfiData)
baseDir <- system.file("extdata", package = "minfiData")
targets <- read.450k.sheet(baseDir)
head(targets)
@
The phenotype data included in the targets file (here \texttt{person, age, sex} and \texttt{status}) will be default phenotype data used in \texttt{shinyMethyl}. We can now extract the necessary data for \texttt{shinyMethyl}:
<<dataExtraction>>=
extractedData <- extractFromTargets450k(targets)
names(extractedData)
@
The extracted data contains different summary statistics of the samples: quantiles of the methylated and unmethylated channels, quantiles of the Beta value distributions, of the M-value distributions, kernel densities, intensities of the control probes in the red and green channels, estimates of the copy number, out-of-band controls and median intensities of the methylated and unmethylated channels for the X and Y chromosomes-mapped probes.
An \texttt{Rda} file containing the extracted data will be automatically saved in the current directory. The file is only a few Mbs in size. To change the name under which the file will be saved, a \texttt{string} can be passed to the function:
<<dataExtractionAlternative1, eval=FALSE>>=
extractedData <- extractFromTargets450k(RGSet, file = "myFileName.Rda")
@
In the case the sample size is large, and not all samples can be loaded into memory, the option \texttt{bySubsets} set to \texttt{TRUE} will modify the function so that only 10 samples are loaded into memory at a time:
<<dataExtractionAlternative2, eval=FALSE>>=
extractedData <- extractFromTargets450k(targets, bySubsets = TRUE)
@
\subsection{Extraction from an \texttt{RGChannelSet}}
An \texttt{RGChannelSet} is a \texttt{minfi} object containing the raw intensities of the green and red channels of the samples. See the \texttt{minfi}'s vignette for more information. Here is an example of how construction such a set with the \texttt{minfi} package from the IDAT files contained in \texttt{minfiData}:
<<RGSetConstruction, eval=TRUE>>=
library(minfiData)
library(minfi)
baseDir <- system.file("extdata", package = "minfiData")
targets <- read.450k.sheet(baseDir)
targets$Sample_Plate <- substr(targets$Slide,1,7)
RGSet <- read.450k.exp(base = baseDir, targets = targets)
@
The phenotype data that will be used in \texttt{shinyMethyl} can be accessed by
<<phenotype, eval=TRUE>>=
pd <- pData(RGSet)
head(pd)
@ 
The function \texttt{extractFromRGSet450k}  is used to extract directly the data from an \texttt{RGChannelSet}:
<<dataExtractionRGSet, eval=FALSE>>=
extractedData <- extractFromRGSet450k(RGSet)
names(extractedData)
@
\section{\texttt{shinyMethyl} session launching}
To launch a \texttt{shinyMethyl} session, simply type the following command:
<<sessionLaunching, eval=FALSE>>=
runShinyMethyl(extractedData)
@


\begin{thebibliography}{9} 
 \bibitem{shiny} RStudio, Inc. 
   \emph{shiny: Web Application Framework for R}, 
   http://cran.r-project.org/web/packages/shiny/index.html
   
    \bibitem{minfi} Kasper Daniel Hansen and Martin Aryee
   \emph{minfi: Analyze
  Illumina's 450k methylation arrays}, 
   R package version 1.7.8
   
     \bibitem{toga} The Cancer Genome Atlas (TCGA)
   \emph{Head and Neck Squamous Cell Carcinoma}, 
   http://cancergenome.nih.gov/cancersselected/headandneck


\end{thebibliography} 

\end{document}